sdk/go/pulumi-language-go/main.go

```diff
@@ -46,36 +53,65 @@ import (
 	pulumirpc "github.com/pulumi/pulumi/sdk/v3/proto/go"
 )
 
-func findProgram(binary string) (*exec.Cmd, error) {
-	// we default to execution via `go run`
-	// the user can explicitly opt in to using a binary executable by specifying
-	// runtime.options.binary in the Pulumi.yaml
-	if binary != "" {
-		program, err := executable.FindExecutable(binary)
+func hashDir(path string) string {
+	contentHash := crypto.SHA256.New()
+	err := filepath.Walk(path, func(path string, info fs.FileInfo, err error) error {
+		if info.IsDir() {
+			// directories do not affect the hash of the program.
+			return nil
+		}
+		data, err := ioutil.ReadFile(path)
 		if err != nil {
-			return nil, errors.Wrap(err, "expected to find prebuilt executable")
+			// unreadable files do not affect the hash of the program.
+			return nil
 		}
-		return exec.Command(program), nil
-	}
+		contentHash.Write([]byte(path))
+		contentHash.Write(data)
+		return nil
+	})
+	contract.IgnoreError(err)
 
-	// Fall back to 'go run' style executions
-	logging.V(5).Infof("No prebuilt executable specified, attempting invocation via 'go run'")
-	program, err := executable.FindExecutable("go")
-	if err != nil {
-		return nil, errors.Wrap(err, "problem executing program (could not run language executor)")
-	}
+	return hex.EncodeToString(contentHash.Sum([]byte{}))
+}
 
+func compileProgramCwd(buildID string) (string, error) {
 	cwd, err := os.Getwd()
 	if err != nil {
-		return nil, errors.Wrap(err, "unable to get current working directory")
+		return "", errors.Wrap(err, "unable to get current working directory")
 	}
 
 	goFileSearchPattern := filepath.Join(cwd, "*.go")
 	if matches, err := filepath.Glob(goFileSearchPattern); err != nil || len(matches) == 0 {
-		return nil, errors.Errorf("Failed to find go files for 'go run' matching %s", goFileSearchPattern)
+		return "", errors.Errorf("Failed to find go files for 'go build' matching %s", goFileSearchPattern)
+	}
+
+	if err != nil {
+		return "", errors.Wrap(err, "failed to compile")
 	}
 
-	return exec.Command(program, "run", cwd), nil
+	hash := hashDir(cwd)
```

Comment:
I would rather we don't try to be smart here for the first pass. Just write to a fresh file in temp, we're not going to be any worse perf wise as that's all "go run" does.

Content hashing like this requires us to understand _all_ content that could possibly feed into the go compilers build process. That includes things like referenced go modules which *do not* live under the source tree of program being built.

Reply:
This is just for naming the output binary destination. The compiler will overwrite the existing binary if it needs to.

cc @t0yv0

Reply:
So don't waste proc time building a hash, just let the temp file API do its job of returning an unused name.